docs/spec/connection-handshake/interfaces.md

```diff
@@ -81,125 +85,58 @@ interface Proof {
 }
 ```
 
-##### New Handler version
-
-Now we redefine `connOpenTry` handler:
+##### New Handler (version 2)
 
+Now we redefine `connOpenTry` handler as follows:
 
 ```typescript
 function connOpenTry(
-  localConnEnd: ConnectionEnd,
-  remoteConnEnd: ConnectionEnd,
+  e: Connection,
   proofInit: Proof,
   proofConsensus: Proof) {
-    localConnId := remoteConnEnd.counterpartyConnectionIdentifier
-    abortTransactionUnless(validateConnectionIdentifier(localConnId))
+    abortTransactionUnless(validateConnectionIdentifier(e.remoteParty.connectionIdentifier))
     abortTransactionUnless(proofConsensus.height <= getCurrentHeight())
     expectedConsensusState = getConsensusState(proofConsensus.height)
 
-    // basic validation of connection ends
-    abortTransactionUnless(remoteConnEnd.state === INIT && localConnEnd.state === INIT)
-    abortTransactionUnless(remoteConnEnd.counterpartyPrefix === getCommitmentPrefix())
+    // basic validation of connection parties
+    abortTransactionUnless(e.state === INIT)
+    abortTransactionUnless(e.remoteParty.prefix === getCommitmentPrefix())
 
-    // verify the proofs
-    abortTransactionUnless(localConnEnd.verifyConnectionState(proofInit,  remoteConnEnd))
-    abortTransactionUnless(localConnEnd.verifyClientConsensusState(
+    abortTransactionUnless(e.remoteParty.verifyConnectionState(proofInit, e))
+    abortTransactionUnless(e.remoteParty.verifyClientConsensusState(
       proofConsensus, proofConsensus.height, expectedConsensusState))
-    previous = provableStore.get(connectionPath(localConnId))
+    previous = provableStore.get(connectionPath(e.remoteParty.connectionIdentifier))
     abortTransactionUnless(
       (previous === null) ||
-      (previous.state === INIT && previous === localConnEnd))
+      (previous === e))
```

Comment:
So this is a bit tricky wrt to the state. It happens that it works here. But in ConnOpenAck the remote state is expected to be TRYOPEN while the local could be INIT or TRYOPEN. If it is INIT this will fail. This is again because e.state is the remote state while previous.state is the local state.

Reply:
I don't think that previous should be equal to e. They should correspond to the same connection, but each from its view. So e.localParty == previous.remoteParty, and vice versa.

Reply:
Note that in this version there is no flipping of _remote_ with _local_. So the module executing this handler sees everything in the `connection` object (and stores it) without flipping any field. For this reason, we can simply do equality check.